ttnn/cpp/ttnn/tensor/xtensor/conversion_utils.hpp

```diff
@@ -0,0 +1,64 @@
+// SPDX-FileCopyrightText: © 2024 Tenstorrent Inc.
+//
+// SPDX-License-Identifier: Apache-2.0
+
+#pragma once
+
+#include "ttnn/tensor/shape/small_vector.hpp"
+#include "ttnn/tensor/tensor.hpp"
+#include <ttnn/tensor/xtensor/xtensor_all_includes.hpp>
+
+namespace ttnn::experimental::xtensor {
+
+// Returns the shape of the xtensor as `ttnn::SimpleShape`.
+template <typename E>
+ttnn::SimpleShape get_shape_from_xarray(const E& xarr) {
+    ttnn::SmallVector<uint32_t> shape_dims;
+    for (size_t i = 0; i < xarr.shape().size(); ++i) {
+        shape_dims.push_back(xarr.shape()[i]);
+    }
+    return ttnn::SimpleShape(shape_dims);
+}
+
+// Converts a span to an xtensor view.
+// IMPORTANT: the lifetime of the returned xtensor view is tied to the lifetime of the underlying buffer.
+template <typename T>
+xt::xarray<T> span_to_xtensor_view(tt::stl::Span<const T> buffer, const ttnn::SimpleShape& shape) {
+    std::vector<size_t> shape_vec(shape.cbegin(), shape.cend());
+    return xt::adapt(buffer.data(), buffer.size(), xt::no_ownership(), shape_vec);
+}
+
+// Converts a span to an xtensor view.
+// IMPORTANT: the lifetime of the returned xtensor view is tied to the lifetime of the underlying buffer.
+template <typename T>
+xt::xarray<T> span_to_xtensor_view(std::span<T> buffer, const ttnn::SimpleShape& shape) {
+    std::vector<size_t> shape_vec(shape.cbegin(), shape.cend());
+    return xt::adapt(buffer.data(), buffer.size(), xt::no_ownership(), shape_vec);
+}
+
+// Converts an xtensor to a span.
+// IMPORTANT: the lifetime of the returned span is tied to the lifetime of the underlying xtensor.
+template <typename T>
+auto xtensor_to_span(const xt::xarray<T>& xtensor) {
+    auto adaptor = xt::adapt(xtensor.data(), xtensor.size(), xt::no_ownership());
+    return tt::stl::Span<const T>(adaptor.data(), adaptor.size());
+}
+
+// Converts an xtensor to a Tensor.
+template <typename T>
+tt::tt_metal::Tensor from_xtensor(const xt::xarray<T>& buffer, const TensorSpec& spec) {
+    auto shape = get_shape_from_xarray(buffer);
+    TT_FATAL(shape == spec.logical_shape(), "xtensor has a different shape than the supplied TensorSpec");
+    auto buffer_view = xtensor_to_span(buffer);
+    return tt::tt_metal::Tensor::from_span<T>(buffer_view, spec);
+}
+
+// Converts a Tensor to an xtensor.
```

Comment:
important - this can be a really expensive operation

Reply:
Added a comment.